Fix some class names

aeszter opened the pull request:
Probably refactoring artifacts.

aeszter committed:
Fix some class names

Probably refactoring artifacts.

iwanders commented:
Hi Aeszter,

Thanks for taking the time to look into this! This is an interesting find... How did you notice this if I may ask? Did you encounter a bug that led you to this or did you just notice it while browsing through the code?

I do think this code actually does what it is supposed to do, but I completely agree it is less than clear. Let us consider the (almost) minimal example which should help explain what happens and why this works (you can run this example to confirm it):
```python
#!/usr/bin/env python3

class Base():
    def __init__(self):
        print("{: <20} {:x}".format("Base", id(self)))

class A(Base):
    def __init__(self):
        print("{: <20} {:x}".format("A", id(self)))
        super(Communicator, self).__init__()

# try importing USB modules here... then we assign make an assignment
Communicator = A

class Recorder(Communicator):
    def __init__(self):
        print("{: <20} {:x}".format("Recorder", id(self)))
        super().__init__()

if __name__ == "__main__":
    instance = A()
    print("")
    record_instance = Recorder()
```

Now it's hopefully already somewhat clear what happens: We define the `Base`, then we define the `A` class (which is `CommunicatorPyUSB` or `CommunicatorHIDAPI` in the real code), in the constructor of this class we use the not-yet-defined `Communicator`. Then, after those definitions we import USB libraries and see which class we would like to use, we assign this to the `Communicator` variable which is then used by the `Recorder` to subclass from, this ensures the recorder can call it's super class (the USB communicator of appropriate type). This prevents having to make two different `Recorder` classes or having to wrap the USB communicator. If the constructor of the `A` class is now called, the `Communicator` variable actually exists in the scope and can be used. Since we either use the `CommunicatorPyUSB` or `CommunicatorHIDAPI` class the `Communicator` will always resolve to the correct type.

So, this works... but I completely agree that it is not clear. For clarity I propose fixing it for the `CommunicatorPyUSB` and `CommunicatorHIDAPI` classes.

We cannot fix this for the `RecorderCommunicator`, since this needs to subclass from the appropriate USB communicator.  The `OfflineCommunicator` could subclass from `BaseCommunicator`, but only if the `__enter__` and `__exit__` methods are also implemented, since those aren't in the base class nor in the `OfflineCommunicator` at the moment.

I hope this clarifies what happens here! You are completely right that this less then clear code was the result of a refactor, majority related to this in 77fb7e31fa6561ecd1618a616efe44ca227642c5. That was a bit of a rush job before my holiday when I had to make this work OS X. I am pretty sure the code works as we used the current version to retrieve logs from two pods during that trip :)

If you update the PR to just replace the `Communicator` variable in the constructors of `CommunicatorPyUSB` and `CommunicatorHIDAPI` I'd be happy to accept it!

aeszter commented:
No, I wouldn't have noticed this merely by looking through the code. My Python-Fu is much too weak. Instead I got a stack trace complaining about Communicator, and I cobbled together this fix from my meager Python skills plus OOP knowledge from other languages.
At the time, I couldn't comprehend how the code had made it into the repo since the problem seemed impossible to miss.
If I rollback now, I cannot reproduce the problem, so I think there must have been another problem. Maybe a missing module that lead to a misleading error message?

aeszter committed:
Partially revoke

"We cannot fix this for the RecorderCommunicator, since this needs to
subclass from the appropriate USB communicator. The
OfflineCommunicator could subclass from BaseCommunicator, but only if
the __enter__ and __exit__ methods are also implemented, since those
aren't in the base class nor in the OfflineCommunicator at the
moment." (iwanders)

Also change CommunicatorHIDAPI along the same lines as
CommunicatorPyUSB.

iwanders commented:
> If I rollback now, I cannot reproduce the problem, so I think there must have been another problem. Maybe a missing module that lead to a misleading error message?

Most certainly! If the usb modules are missing all you get is a lousy print statement... This is fine in my opinion as you can still use the code on the filesystem dumps for retrieval / analysis etc. The `requirements.txt` and elaborate documentation should prevent anyone from using it without the usb modules installed. Perhaps the error could be more clear though.

Thanks for updating the PR and helping clean up the code! :+1:
